server/src/main/java/org/elasticsearch/index/mapper/AnnotatedTextFieldMapper.java

```diff
@@ -0,0 +1,815 @@
+import org.apache.lucene.analysis.TokenFilter;
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.Tokenizer;
+import org.apache.lucene.analysis.ngram.EdgeNGramTokenFilter;
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.OffsetAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionLengthAttribute;
+import org.apache.lucene.analysis.tokenattributes.TypeAttribute;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.index.IndexOptions;
+import org.apache.lucene.index.IndexableField;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.MultiTermQuery;
+import org.apache.lucene.search.NormsFieldExistsQuery;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.TermQuery;
+import org.elasticsearch.ElasticsearchParseException;
+import org.elasticsearch.common.collect.Iterators;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.index.analysis.AnalyzerScope;
+import org.elasticsearch.index.analysis.NamedAnalyzer;
+import org.elasticsearch.index.fielddata.IndexFieldData;
+import org.elasticsearch.index.fielddata.plain.PagedBytesIndexFieldData;
+import org.elasticsearch.index.mapper.AnnotatedTextFieldMapper.AnnotatedText.AnnotationToken;
+import org.elasticsearch.index.mapper.TextFieldMapper.PrefixFieldType;
+import org.elasticsearch.index.mapper.TextFieldMapper.PrefixFieldMapper;
+
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.io.Reader;
+import java.io.StringReader;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static org.elasticsearch.index.mapper.TypeParsers.parseTextField;
+
+/** A {@link FieldMapper} for full-text fields with annotation markup e.g.
+ * 
+ *    "New mayor is [John Smith](type=person&amp;value=John%20Smith) "
+ * 
+ * A special Analyzer wraps the default choice of analyzer in order
+ * to strip the text field of annotation markup and inject the related
+ * entity annotation tokens as supplementary tokens at the relevant points
+ * in the token stream.
+ * This code is largely a copy of TextFieldMapper which is less than ideal - 
+ * my attempts to subclass TextFieldMapper faild but we can revisit this.
+ **/
+public class AnnotatedTextFieldMapper extends FieldMapper {
```

Comment:
To simplify the usage of this field type I think we should automatically (or add an option to activate it) index the annotated text in a `doc_values` field. This field would be available for aggregations and the terms bucket can be used to generate queries for highlighting or matching. I know it's possible to just index the annotations in a different field but I think it would be beneficial to have to link these two fields internally.

Reply:
There's a number of questions around that:
1) What if users want the source JSON to include a structured "person" array? 
2) How do we define the "copy_to" fields in the annotation markup?
3) Are there any complexities if the annotated text field appears under a `nested` object?

Could we save that feature for a phase 2 and keep this initial PR simpler for the moment?

Reply:
We have additional fields for other use cases (parent/child, prefix text, ...) so it can work here too. The name of the field should be derived from the main field name, it can be hardcoded or the suffix can be added as an option in the field. 
We can add this feature to phase 2 but I thought that it would be easier to have a clear link between the two fields in the first iteration. Let's leave this for a follow up.